opcpa-tpr-config/opcpa-tpr-config.py

```diff
@@ -0,0 +1,164 @@
+                        pv = f"ca://{tpr_base}:CH{tpr_ch}_{label}"
+                        child.set_channel(pv)
+                    grid.addWidget(child, nchannel, nlabel)
+        else:
+            print("Laser {} not found!".format(laser))
+
+    def setup_configs(self, laser):
+        """
+        Setup the configuration buttons for the given laser system.
+
+        Arguments
+        ---------
+        laser: The name of the laser configuration to be used.
+        """
+        grid = self.findChild(QGridLayout, f"{laser}_config_layout")
+        if grid is not None:
+            las = self.config["lasers"][laser]
+            for ncfg, cfg in enumerate(self.config[las]["rate_configs"]):
+                button = PyDMPushButton(f"{laser}_{cfg}")
+                rate = self.config[las]["rate_configs"][cfg]["rate"]
+                button.setText(f"{rate}")
+                button.showConfirmDialog = True
+                row, col = divmod(ncfg, 4)
+                row += 1  # Leave first row for title
+                grid.addWidget(button, row, col)
+                button.clicked.connect(partial(self.set_configuration, las, cfg))
+
+    def set_configuration(self, laser, config):
+        """
+        Apply the given configuration to the TPR.
+
+        Arguments
+        ---------
+        laser: The name of the laser configuration to be used.
+
+        config: The rep rate configuration to be applied when calling this
+                function.
+        """
+        tpr_base = self.config[laser]["tpr_base"]
+        rate_conf = self.config[laser]["rate_configs"][config]
+        for channel in self.config[laser]["channels"]:
+            if channel in self.config[laser]["rate_configs"][config]:
+                tpr_ch = self.config[laser]["channels"][f"{channel}"]["ch"]
+                # Set rate mode
+                ratemode_val = rate_conf[channel]["ratemode"]
+                ratemode_sig = EpicsSignal(f"{tpr_base}:CH{tpr_ch}_RATEMODE")
+                ratemode_sig.put(ratemode_val)
+                # Set rate
+                rate_val = rate_conf[channel]["rate"]
+                if ratemode_val == "Fixed":
+                    rate_sig = EpicsSignal(f"{tpr_base}:CH{tpr_ch}_FIXEDRATE")
+                    rate_sig.put(rate_val)
+                elif ratemode_val == "Seq":
+                    rate_sig = EpicsSignal(f"{tpr_base}:CH{tpr_ch}_SEQCODE")
+                    event_code = self.config[laser]["rep_rates"][rate_val]
+                    rate_sig.put(event_code)
+                else:
+                    raise ValueError(f"Unknown ratemode {ratemode_val}")
+                # Enable/Disable the trigger
+                enable_val = rate_conf[channel]["enable"]
+                trg_sig = EpicsSignal(f"{tpr_base}:TRG{tpr_ch}_SYS2_TCTL")
+                trg_sig.put(enable_val)
+                ch_sig = EpicsSignal(f"{tpr_base}:CH{tpr_ch}_SYS2_TCTL")
+                ch_sig.put(enable_val)
```

Comment:
Maybe a super minor detail: if you make all the `EpicsSignal` objects first, they will connect in the background, reducing the runtime of each `put` method call later, which needs to wait for connection.

Reply:
Good to know, I'll update the order

Reply:
On second thought maybe I won't because I don't like how it looks, and the run time performance was near instant. But it's a good thing to know for other applications.